drivers/delta-hid.c

```diff
@@ -0,0 +1,188 @@
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ */
+
+#include "usbhid-ups.h"
+#include "delta-hid.h"
+#include "main.h"	/* for getval() */
+#include "usb-common.h"
+
+#define DELTA_HID_VERSION	"Delta HID 0.5"
+
+/* Delta Vendor ID */
+#define DELTA_VENDORID	0x05dd
+
+/* USB IDs device table */
+static usb_device_id_t delta_usb_device_table[] = {
+	/* Delta RT Series, Single Phase, 1/2/3 kVA */
+	{ USB_DEVICE(DELTA_VENDORID, 0x041b), NULL },
+
+	/* Terminating entry */
+	{ -1, -1, NULL }
+};
+
+
+/* --------------------------------------------------------------- */
+/*      Vendor-specific usage table */
+/* --------------------------------------------------------------- */
+
+/* DELTA usage table */
+static usage_lkp_t delta_usage_lkp[] = {
+	{ "DELTA1",	0x00000000 },
+	{ "DELTA2",	0xff000055 },
+	{ "DELTA3",	0xffff0010 },
+	{ "DELTA4",	0xffff0056 },
+	{ "DELTA5",	0xffff0057 },
+	{ "DELTA6",	0xffff0058 },
+	{ "DELTA7",	0xffff0059 },
+	{ "DELTA8",	0xffff005a },
+	{ "DELTA9",	0xffff0060 },
+	{ "DELTA10",	0xffff0061 },
+	{ "DELTA11",	0xffff0062 },
+	{ "DELTA12",	0xffff0063 },
+	{ "DELTA13",	0xffff0064 },
+	{ "DELTA14",	0xffff0065 },
+	{ "DELTA15",	0xffff0066 },
+	{ "DELTA16",	0xffff0067 },
+	{ "DELTA17",	0xffff0068 },
+	{ "DELTA18",	0xffff0075 },
+	{ "DELTA19",	0xffff0076 },
+	{ "DELTA20",	0xffff007c },
+	{ "DELTA21",	0xffff007d },
+	{ "DELTA22",	0xffff0081 },
+	{ "DELTA23",	0xffff0091 },
+	{ "DELTA24",	0xffff0092 },
+	{ "DELTA25",	0xffff0093 },
+	{ "DELTA26",	0xffff0094 },
+	{ "DELTA27",	0xffff0095 },
+	{ "DELTA28",	0xffff0096 },
+	{ "DELTA29",	0xffff0097 },
+	{ "DELTA30",	0xffff0098 },
+	{ "DELTA31",	0xffff0099 },
+	{ "DELTA32",	0xffff009a },
+	{ "DELTA33",	0xffff009b },
+	{ "DELTA34",	0xffff009c },
+	{  NULL, 0 }
+};
+
+static usage_tables_t delta_utab[] = {
+	delta_usage_lkp,
+	hid_usage_lkp,
+	NULL,
+};
+
+/* --------------------------------------------------------------- */
+/* HID2NUT lookup table                                            */
+/* --------------------------------------------------------------- */
+
+static hid_info_t delta_hid2nut[] = {
+
+  { "battery.voltage.nominal", 0, 0, "UPS.BatterySystem.Battery.ConfigVoltage", NULL, "%.0f", HU_FLAG_STATIC, NULL },
+  { "battery.voltage", 0, 0, "UPS.BatterySystem.Battery.Voltage", NULL, "%.1f", HU_FLAG_QUICK_POLL, NULL },
```

Comment:
HU_FLAG_QUICK_POLL should be limited to status data and closely monitored ones, such as timers , that need refresh more often. Just put 0 for other vars

Reply:
Polling Battery.Voltage once in minute - not good, under high load battery may discharge really fast within minute.

Reply:
I suppose that would mean that the UPS is under-spec'ed for its load. The battery capacity should allow for some overhead of detection of a non-negligibly long outage, and then some for wall power to maybe come back quickly, and when the battery level gets to suspicious levels - initiate graceful shutdown of systems and survive long enough (deployment-dependent, servers can take tens of minutes for interconnected webs of services to go down correctly).

And also batteries age, so it might not hold as long in practice when the outage bites as when it did when you bought and tested it.

Rule of thumb is that you should have at least 15 minutes under your full load (e.g. server busy checksumming its disks over and over with all cores and disks busy).

So if the "once a minute" is too infrequent, something is wrong on the hardware side.

Also, voltage the battery produces is often non-linear compared to its remaining capacity or run-time; if the UPS hardware and protocol provides *those* data points, and does not mislead doing so (I think we had an issue open for some devices that publish unreliable values), they should be preferred for determining when you should start a shutdown.

That all ranted out, @aquette : in case of going on-battery, 1) Does an USB UPS actively push that notice into the driver or is it only seen with latency of regular polling? 2) Does the poll frequency increase when we are at risk?

Reply:
1) supposed to push notifications, but really depends on the mfr implem. Visible in the initial dump, if this var is both tagged notif and polled (don't remember the wording I put though)
2) no, but could be an improvement to consider. Still, some device may not support well too much polling when on batt, since they are busy processing internals. And still, interrupts notif are supposed to be here for that!